linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,516 @@
+#include "c_pal/sync.h"
+#include "c_pal/srw_lock.h"
+#include "c_pal/execution_engine.h"
+#include "c_pal/execution_engine_linux.h"
+
+#include "c_pal/threadpool.h"
+
+#define TIMEOUT_MS          100
+#define MAX_THREAD_COUNT    1
+#define THREAD_LIMIT        20
+
+#define TASK_RESULT_VALUES  \
+    TASK_NOT_USED,          \
+    TASK_INITIALIZING,      \
+    TASK_WAITING,           \
+    TASK_WORKING
+
+MU_DEFINE_ENUM_WITHOUT_INVALID(TASK_RESULT, TASK_RESULT_VALUES);
+MU_DEFINE_ENUM_STRINGS(TASK_RESULT, TASK_RESULT_VALUES)
+
+#define THREADPOOL_STATE_VALUES \
+    THREADPOOL_STATE_NOT_OPEN,  \
+    THREADPOOL_STATE_OPENING,   \
+    THREADPOOL_STATE_OPEN,      \
+    THREADPOOL_STATE_CLOSING
+
+MU_DEFINE_ENUM(THREADPOOL_STATE, THREADPOOL_STATE_VALUES)
+
+MU_DEFINE_ENUM_STRINGS(THREADPOOL_STATE, THREADPOOL_STATE_VALUES)
+
+typedef struct THREADPOOL_TASK_TAG
+{
+    volatile_atomic int32_t task_state;
+    THREADPOOL_WORK_FUNCTION task_func;
+    void* task_param;
+} THREADPOOL_TASK;
+
+typedef struct THREADPOOL_TAG
+{
+    volatile_atomic int32_t state;
+    volatile_atomic int32_t thread_count;
+    uint32_t max_thread_count;
+    uint32_t min_thread_count;
+    int32_t used_thread_count;
+    volatile_atomic int32_t quitting;
+    volatile_atomic int32_t task_count;
+
+    sem_t semaphore;
+
+    // Pool Array
+    THREADPOOL_TASK* task_array;
+    volatile_atomic int64_t insert_idx;
+    volatile_atomic int64_t consume_idx;
+    volatile_atomic int32_t task_array_size;
+    volatile_atomic int32_t pending_call_count;
+
+    SRW_LOCK_HANDLE srw_lock;
+    uint32_t list_index;
+
+    THREAD_HANDLE* thread_handle_array;
+} THREADPOOL;
+
+static void internal_close(THREADPOOL_HANDLE threadpool)
+{
+    bool should_wait_for_transition = false;
+    do
+    {
+        THREADPOOL_STATE current_state = interlocked_compare_exchange(&threadpool->state, THREADPOOL_STATE_CLOSING, THREADPOOL_STATE_OPEN);
+        if (current_state == THREADPOOL_STATE_NOT_OPEN || current_state == THREADPOOL_STATE_CLOSING)
+        {
+            should_wait_for_transition = false;
+        }
+        else if (current_state == THREADPOOL_STATE_OPENING)
+        {
+            should_wait_for_transition = true;
+        }
+        else
+        {
+            (void)interlocked_exchange(&threadpool->quitting, 1);
+
+            while (interlocked_add(&threadpool->pending_call_count, 0) != 0)
+            {
+                ThreadAPI_Sleep(1);
+            }
+
+            for (int32_t index = 0; index < threadpool->used_thread_count; index++)
+            {
+                int dont_care;
+                if (ThreadAPI_Join(threadpool->thread_handle_array[index], &dont_care) != THREADAPI_OK)
+                {
+                    LogError("Failure joining thread number %" PRId32 "", index);
+                }
+            }
+            (void)interlocked_exchange(&threadpool->state, THREADPOOL_STATE_NOT_OPEN);
+            should_wait_for_transition = true;
```

Comment:
>hould_wait_for_transition = true; [](http://example.com/codeflow?start=13&length=33)

seems like this does an extra loop iteration just to check the state again, I'd expect it to be `false` instead to break out of the loop.